### DISCUSSION TITLE
Change the CLR Type, without changing all resolvery using it

### DISCUSSION THREAD

**Discussion by lanwin** [graphql-dotnet/graphql-dotnet#4066] [Q&A] [answered]
Hi, I have a large app running graphql-dotnet and want to wrap one type in a context, but without changing all the type usage (since this are a lot of places).

What I have:
```
// Model
class Foo{}
// Type
class FooType : ObjectGraphType<Foo>
```

What I want:
```
// Model
class Foo{}
//Context
class FooContext(Foo Foo);
// Type
class FooType : ObjectGraphType<FooContext>
```

Now all resolvers currently returning `Foo` are failing.

What I want is that Graphql-Dotnet should convert `Foo` to `FooContext` behind the scenes and use `FooType` with that.

I tried many different things like 

* schema.RegisterTypeMapping(typeof(Foo),typeof(FooType));
* add ValueConverters in both directions
* add implictly conversion operators

But non of them make it work. Is there any way?

**Comment by Shane32** [comment, marked as answer]
You could use field middleware to convert Foo objects to FooContext.  Can I ask why you would want this?  There might be another way to do what you're trying to accomplish.

Here's a ChatGPT description of this process, which is probably mostly correct:

---

To implement a middleware that converts `Foo` objects into `FooContext` objects in GraphQL.NET, you need to implement the `IFieldMiddleware` interface and intercept the field resolution process. In this case, whenever a field resolves to a `Foo` object, the middleware will wrap it in a `FooContext` object.

Here's a sample implementation of this:

```csharp
using GraphQL;
using GraphQL.Resolvers;
using GraphQL.Types;
using System;
using System.Threading.Tasks;

// Define your Foo and FooContext classes
public class Foo
{
    // Foo properties
}

public class FooContext
{
    public Foo FooValue { get; }

    public FooContext(Foo foo)
    {
        FooValue = foo;
    }
}

// Implement the middleware
public class FooToFooContextMiddleware : IFieldMiddleware
{
    public async ValueTask<object?> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
    {
        // First, let the execution go to the next middleware or field resolver
        var result = await next(context);

        // Check if the result is a Foo object and convert it to FooContext
        if (result is Foo foo)
        {
            return new FooContext(foo);
        }

        // Otherwise, return the result as-is
        return result;
    }
}

// Register the middleware
public class MySchema : Schema
{
    public MySchema(IServiceProvider provider) : base(provider)
    {
        // Add middleware to the schema, this applies the middleware globally
        this.FieldMiddleware.Use(new FooToFooContextMiddleware());
    }
}
```

### Explanation:

1. **Foo and FooContext**: These are your classes where `FooContext` wraps the `Foo` object.

2. **FooToFooContextMiddleware**: This is the actual middleware implementation. It implements `IFieldMiddleware` and overrides the `Resolve` method. Inside this method:
   - The `next(context)` delegate is invoked to resolve the field, potentially returning a `Foo` object.
   - If the result is a `Foo` object, it is wrapped in a `FooContext` object.
   - If the result is something else, it's returned as-is.

3. **Schema Registration**: The middleware is registered in the schema via `this.FieldMiddleware.Use(new FooToFooContextMiddleware())`. This ensures the middleware applies globally to all field resolutions.

This middleware can be added to any GraphQL.NET project and will automatically wrap any `Foo` results in `FooContext`.

**Reply by Shane32** [reply]
Note that this solution will execute for every field resolver in the schema, not just resolvers that return Foo.  So there is some overhead involved, but it should work.

**Reply by lanwin** [reply]
Thank you for your answer! I did not thought that ChatGPT would give a good answer here...but lessen learned.

But you asked for the reason why I want to do this. 

The Foo object is a Database object and currently it basically always returns the some properties for the same Foo. But we want to change that to return different values on different context's.

In example imagine the FooType has a text property. Currently that property returns simply the text of that object. Now we have different context's where we need to reduce that text to a (lets say copyright safe) version depending on the context. There are also some links like a pdfUrl property (witch returns a pdf link for that object) witch should resolve to different destinations depending on the context in that the object is used. The object itself is still a FooType and should work with all code that accepts Foo object on the client.

The problem is that this context is not globally. In one part of the graph it could return the same Foo object with context A and on the other part of the graph it can be context B. 

So my idea was to change the FooType to use a FooContext where I can transport down some information from the context to the resolvers of FooType. Most of the time that is the default context. This is why I can live without changing every usage of Foo to FooContext. I also dont want to use interfaces here, cause this object is large and most of the properties are exactly the same in all context's. Its also exactly the same object. So using multiple types for each context does not seems to be right here.

**Reply by lanwin** [reply]
Sadly this solution did not seem to work. I get `System.InvalidCastException: 'Unable to cast object of type 'Model.Foo' to type 'Model.FooContext'.'` on `var result = await next(context);`

**Reply by Shane32** [reply]
If you can post a simple project that reproduces the issue, I can debug. And/or if you can post the full stack trace at the InvalidCastException that would also help.

**Reply by Shane32** [reply]
> Thank you for your answer! I did not thought that ChatGPT would give a good answer here...but lessen learned.

To be fair, I didn't actually ask it to solve the problem, but just prompted it to write a sample, as follows:

> write an IFieldMiddleware implementation for GraphQL.NET that converts any Foo objects to FooContext objects where FooContext has a constructor of FooContext(Foo value)

It usually writes code that favors older versions of GraphQL.NET, and in this case I had to change the IFieldMiddleware delegate to return a `ValueTask<object?>` instead of `Task<object>` but otherwise its response was what I was looking for, even if a little long-winded.  I see now that `Resolve` should have been changed to `ResolveAsync` also.  Again, it's trained on an older version of GraphQL.NET.

**Reply by Shane32** [reply]
So now to write a test, I prompt for this:

> write a sample schema using that middleware where one field returns Foo and another FooContext.  Add a property to Foo for the sake of the sample.  Complete the sample with an XUnit test to execute a request against the sample schema, using Shouldly to verify the actual JSON serialized response against the expected JSON response.

And it gives something fairly close, and with a little work I have a POC that passes tests.

```cs
using GraphQL.Instrumentation;
using GraphQL.Types;

namespace GraphQL.Tests.Bugs;

public class Bug4066
{
    public class Foo
    {
        public string Name { get; set; }
    }

    public class FooContext
    {
        public Foo FooValue { get; }

        public FooContext(Foo foo)
        {
            FooValue = foo;
        }
    }

    // Define the Query Type
    public class SampleQuery : ObjectGraphType
    {
        public SampleQuery()
        {
            // Field that returns a Foo object with an explicit graph type
            Field<FooContextType>("foo").Resolve(_ => new Foo { Name = "Sample Foo" });

            // Field that returns a FooContext object with an explicit graph type
            Field<FooContextType>("fooContext").Resolve(_ => new FooContext(new Foo { Name = "Contextual Foo" }));

            // Field that returns a Foo object and infers the graph type via registered CLR type mappings
            Field("foo2", _ => new Foo { Name = "Sample Foo 2" });

            // Field that returns a FooContext object and infers the graph type via registered CLR type mappings
            Field("fooContext2", _ => new FooContext(new Foo { Name = "Contextual Foo 2" }));
        }
    }

    // Define the FooContext GraphQL type
    public class FooContextType : ObjectGraphType<FooContext>
    {
        public FooContextType()
        {
            Field(x => x.FooValue.Name);
        }
    }

    // Define the Schema
    public class SampleSchema : Schema
    {
        public SampleSchema(IServiceProvider provider) : base(provider)
        {
            Query = new SampleQuery();
            FieldMiddleware.Use(new FooToFooContextMiddleware());
            // need to explicitly register the Foo to FooContextType mapping
            this.RegisterTypeMapping<Foo, FooContextType>();
            // FooContext is implicit when .AddClrTypeMappings() is in use
            this.RegisterTypeMapping<FooContext, FooContextType>();
        }
    }

    // Implement the middleware
    public class FooToFooContextMiddleware : IFieldMiddleware
    {
        public async ValueTask<object?> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
        {
            // First, let the execution go to the next middleware or field resolver
            var result = await next(context);

            // Check if the result is a Foo object and convert it to FooContext
            if (result is Foo foo)
            {
                return new FooContext(foo);
            }

            // Otherwise, return the result as-is
            return result;
        }
    }


    // Create a test method to run the GraphQL query
    [Fact]
    public async Task FooAndFooContext_ShouldReturnExpectedJson()
    {
        // Arrange: Create schema and execute the query
        var schema = new SampleSchema(new DefaultServiceProvider());

        var query = @"
        {
            foo { name }
            fooContext { name }
            foo2 { name }
            fooContext2 { name }
        }"
        ;

        var json = await schema.ExecuteAsync(_ =>
        {
            _.Schema = schema;
            _.Query = query;
        });

        // Assert: Use Shouldly to verify the actual response matches expected response
        var expectedJson = """
            {
                "data": {
                    "foo": {
                        "name": "Sample Foo"
                    },
                    "fooContext": {
                        "name": "Contextual Foo"
                    },
                    "foo2": {
                        "name": "Sample Foo 2"
                    },
                    "fooContext2": {
                        "name": "Contextual Foo 2"
                    }
                }
            }
            """;

        json.ShouldBeCrossPlatJson(expectedJson);
    }
}
```

**Reply by Shane32** [reply]
All set @lanwin ?

**Reply by lanwin** [reply]
Thank you for you time. I had not the chance yet to try it. Maybe Monday or Thursday.

**Reply by lanwin** [reply]
Hi @Shane32, I was able to try it now and it failes for a connection type at `var result = await next(context);` with can not convert from Foo to FooType. The code looked something like this:

```cs
Field("foo", typeof(NonNullGraphType<ConnectionType<FooContextType>>))
    .ResolveAsync(GetFoo);
    
async Task<Connection<Foo>> GetFoo(IResolveFieldContext<Model> context)
{
    return new Connection<Foo>
    {
        Edges = [new Edge<Foo>(){Id=10, Node=new Foo(){Name="Test"} ]
    };
}
```

If I change `Connection<Foo>` to `Connection<FooContext>` it works. Do you see a way to make that work anyway?

**Reply by Shane32** [reply]
try changing this:

```cs
            // Check if the result is a Foo object and convert it to FooContext
            if (result is Foo foo)
            {
                return new FooContext(foo);
            }
```

to:

```cs
            // Check if the result is a Foo object and convert it to FooContext
            if (result is Foo foo)
            {
                return new FooContext(foo);
            }
            else if (result is IEnumerable<Foo> fooList)
            {
                return fooList.Select(x => new FooContext(x));
            }
```

**Reply by lanwin** [reply]
Ok I did not thought that works since I got the error on the lines below. But it really worked. 

Thank you! 

Hope this answer helps someone else too in the future.
